File: VeterinariaLosDogos/app/src/main/java/com/example/myapplication/fragments/ProfileFragment.java
package com.example.myapplication.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.myapplication.R;
import com.example.myapplication.models.IoTStatus;
import com.example.myapplication.models.Pet;
import com.example.myapplication.models.User;
import com.example.myapplication.repository.IoTRepository;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ServerValue;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class ProfileFragment extends Fragment {

    private TextView tvUserName, tvUserEmail;
    private LinearLayout petsContainer;

    private FirebaseAuth mAuth;
    private DatabaseReference db;

    private Query petsQuery;
    private ValueEventListener petsListener;

    public ProfileFragment() {}

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_profile, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View v, @Nullable Bundle savedInstanceState) {
        tvUserName = v.findViewById(R.id.tvUserName);
        tvUserEmail = v.findViewById(R.id.tvUserEmail);
        petsContainer = v.findViewById(R.id.petsContainer);

        mAuth = FirebaseAuth.getInstance();
        db = FirebaseDatabase.getInstance().getReference();

        loadUser();
    }

    private void loadUser() {
        FirebaseUser current = mAuth.getCurrentUser();
        if (current == null) return;

        db.child("users").child(current.getUid())
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override public void onDataChange(@NonNull DataSnapshot snapshot) {
                        User u = snapshot.getValue(User.class);
                        if (u != null) {
                            tvUserName.setText(u.getName());
                            tvUserEmail.setText(u.getEmail());
                            loadPets(current.getUid());
                        }
                    }

                    @Override public void onCancelled(@NonNull DatabaseError error) { }
                });
    }

    private void loadPets(@NonNull String ownerId) {
        petsContainer.removeAllViews();

        petsQuery = db.child("pets")
                .orderByChild("ownerId")
                .equalTo(ownerId);

        petsListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot sn) {
                petsContainer.removeAllViews();
                List<Pet> pets = new ArrayList<>();

                for (DataSnapshot child : sn.getChildren()) {
                    Pet p = child.getValue(Pet.class);
                    if (p != null) pets.add(p);
                }

                for (Pet p : pets) {
                    addPetView(p);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) { }
        };

        petsQuery.addValueEventListener(petsListener);
    }

    private void addPetView(@NonNull Pet p) {
        int pad = (int) (8 * getResources().getDisplayMetrics().density);

        // Nombre de la mascota
        TextView tv = new TextView(getContext());
        tv.setPadding(pad, pad, pad, pad);
        tv.setText("🐾 " + p.getName() + " (" + p.getSpecies() + ")");
        petsContainer.addView(tv);

        // TextView para mostrar estado IoT
        TextView tvIot = new TextView(getContext());
        tvIot.setPadding(pad, 0, pad, pad);
        tvIot.setText("Cargando datos IoT…");
        petsContainer.addView(tvIot);

        // Repositorio IoT
        IoTRepository ioTRepo = new IoTRepository();

        // Escuchar estado IoT SOLO si la mascota tiene id
        if (p.getId() != null) {
            ioTRepo.listenToStatus(p.getId(), new ValueEventListener() {
                @Override
                public void onDataChange(@NonNull DataSnapshot snapshot) {

                    IoTStatus st = snapshot.getValue(IoTStatus.class);
                    if (st == null) {
                        tvIot.setText("Sin datos IoT");
                        return;
                    }

                    tvIot.setText(
                            "Temp: " + st.getTemperature() + "°C\n" +
                                    "FC: " + st.getHeartRate() + " lpm\n" +
                                    "Actividad: " + st.getActivityLevel() + "\n" +
                                    "Puerta: " + (st.isDoorOpen() ? "Abierta" : "Cerrada") + "\n" +
                                    "Buzzer: " + (st.isBuzzerActive() ? "Activo" : "Off") + "\n" +
                                    "Estado: " + st.getState()
                    );
                }

                @Override
                public void onCancelled(@NonNull DatabaseError error) {
                    tvIot.setText("Error al leer IoT: " + error.getMessage());
                }
            });
        } else {
            tvIot.setText("Mascota sin ID, no se puede vincular IoT.");
        }

        // Botón para enviar comando a actuador (buzzer/LED en el microcontrolador)
        Button btn = new Button(getContext());
        btn.setText("Activar alarma IoT");
        btn.setOnClickListener(v -> {
            if (p.getId() != null) {
                sendIotCommand(p.getId(), "ALARM_ON");
            } else {
                Toast.makeText(getContext(), "Mascota sin ID para IoT", Toast.LENGTH_SHORT).show();
            }
        });

        petsContainer.addView(btn);
    }

    // Mantengo este método para enviar comandos IoT (actuadores)
    private void sendIotCommand(@NonNull String petId, @NonNull String command) {
        String cmdId = db.child("iotCommands").child(petId).push().getKey();
        if (cmdId == null) return;

        Map<String, Object> payload = new HashMap<>();
        payload.put("command", command);
        payload.put("timestamp", ServerValue.TIMESTAMP);

        db.child("iotCommands").child(petId).child(cmdId)
                .setValue(payload)
                .addOnSuccessListener(unused -> {
                    Toast.makeText(getContext(), "Comando enviado", Toast.LENGTH_SHORT).show();
                });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (petsQuery != null && petsListener != null)
            petsQuery.removeEventListener(petsListener);
    }
}

File: VeterinariaLosDogos/app/src/main/java/com/example/myapplication/fragments/ProfileFragment.kt
package com.example.myapplication.fragments

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import coil.load
import coil.transform.CircleCropTransformation
import com.example.myapplication.R
import com.example.myapplication.models.DogBreed
import com.example.myapplication.models.Pet
import com.example.myapplication.network.ApiService
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class ProfileFragment : Fragment() {

    private lateinit var auth: FirebaseAuth
    private lateinit var db: DatabaseReference
    private lateinit var petsContainer: LinearLayout
    private lateinit var tvUserName: TextView
    private lateinit var tvUserEmail: TextView
    private lateinit var imgUserProfile: ImageView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_profile, container, false)

        // Inicializar Firebase
        auth = FirebaseAuth.getInstance()
        db = FirebaseDatabase.getInstance().reference

        // Referencias a la UI
        tvUserName = view.findViewById(R.id.tv_profile_name)
        tvUserEmail = view.findViewById(R.id.tv_profile_email)
        imgUserProfile = view.findViewById(R.id.img_user_profile)
        petsContainer = view.findViewById(R.id.pets_container)

        loadUserData()
        loadUserPets()
        fetchDogBreedsSuggestions()

        return view
    }

    // 1. Cargar datos del Usuario desde Firebase
    private fun loadUserData() {
        val userId = auth.currentUser?.uid ?: return
        db.child("Users").child(userId).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val name = snapshot.child("name").getValue(String::class.java)
                val email = snapshot.child("email").getValue(String::class.java)

                tvUserName.text = name ?: "Usuario"
                tvUserEmail.text = email ?: "Sin correo"

                // Aquí podrías cargar la foto de perfil del usuario si ya la tienes en Firebase
                // imgUserProfile.load(snapshot.child("imageUrl").value) { transformations(CircleCropTransformation()) }
            }
            override fun onCancelled(error: DatabaseError) {
                Log.e("Firebase", "Error al cargar usuario", error.toException())
            }
        })
    }

    // 2. Cargar lista de Mascotas con sus Fotos
    private fun loadUserPets() {
        val userId = auth.currentUser?.uid ?: return
        db.child("pets").orderByChild("ownerId").equalTo(userId)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    petsContainer.removeAllViews() // Limpiar antes de recargar
                    for (postSnapshot in snapshot.children) {
                        val pet = postSnapshot.getValue(Pet::class.java)
                        if (pet != null) {
                            addPetToUI(pet)
                        }
                    }
                }
                override fun onCancelled(error: DatabaseError) {}
            })
    }

    private fun addPetToUI(pet: Pet) {
        val petView = layoutInflater.inflate(R.layout.item_pet_card, null)
        val nameTv = petView.findViewById<TextView>(R.id.tv_item_pet_name)
        val breedTv = petView.findViewById<TextView>(R.id.tv_item_pet_breed)
        val petImg = petView.findViewById<ImageView>(R.id.img_item_pet)

        nameTv.text = pet.name
        breedTv.text = pet.breed

        // Cargar imagen de la mascota usando la librería COIL
        if (!pet.imageUrl.isNullOrEmpty()) {
            petImg.load(pet.imageUrl) {
                crossfade(true)
                placeholder(R.drawable.ic_pet_placeholder) // Imagen mientras carga
                transformations(CircleCropTransformation())
            }
        }

        petsContainer.addView(petView)
    }

    // 3. Consumo de API REST con Retrofit (Requerimiento 2.2)
    private fun fetchDogBreedsSuggestions() {
        val retrofit = Retrofit.Builder()
            .baseUrl("https://api.thedogapi.com/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        val service = retrofit.create(ApiService::class.java)
        service.getBreeds().enqueue(object : Callback<List<DogBreed>> {
            override fun onResponse(call: Call<List<DogBreed>>, response: Response<List<DogBreed>>) {
                if (response.isSuccessful) {
                    val breeds = response.body()?.take(3) // Mostrar solo 3 sugerencias
                    breeds?.forEach { breed ->
                        val suggestionTv = TextView(context)
                        suggestionTv.text = "💡 Sugerencia: Conoce al ${breed.name}"
                        suggestionTv.setPadding(10, 10, 10, 10)
                        petsContainer.addView(suggestionTv)
                    }
                }
            }
            override fun onFailure(call: Call<List<DogBreed>>, t: Throwable) {
                Log.e("API", "Error: ${t.message}")
            }
        })
    }
}